File: apps/backend/src/schemas/retrieval.py
"""Retrieval data models and schemas."""

from typing import Literal

from src.schemas.base import CamelModel

RetrievalStatus = Literal["not_ready", "indexing", "ready", "error"]


class RetrievalChunk(CamelModel):
    id: str
    video_id: str = ""
    chunk_index: int = 0
    text: str
    start: float
    end: float
    segment_start_index: int = 0
    segment_end_index: int = 0
    token_count: int = 0


class RetrievalResult(CamelModel):
    chunk: RetrievalChunk
    score: float


class RetrievalRequest(CamelModel):
    query: str
    top_k: int = 5


class RetrievalResponse(CamelModel):
    success: bool = True
    video_id: str
    results: list[RetrievalResult]
    request_id: str | None = None

File: packages/shared-types/src/retrieval.ts
/**
 * Status lifecycle states for video retrieval indexing.
 */
export type RetrievalStatus = "not_ready" | "indexing" | "ready" | "error";

/**
 * A normalized, timestamped retrieval chunk traceable to the original transcript segments.
 * Preserves timing boundaries for citation seeking and downstream LLM grounding.
 */
export interface RetrievalChunk {
  id: string;
  videoId: string;
  chunkIndex: number;
  text: string;
  start: number;
  end: number;
  segmentStartIndex: number;
  segmentEndIndex: number;
  tokenCount: number;
}

/**
 * Scored search result returned by the retriever.
 */
export interface RetrievalResult {
  chunk: RetrievalChunk;
  score: number;
}

/**
 * Request payload to query the video retrieval index.
 */
export interface RetrievalRequest {
  query: string;
  topK?: number;
}

/**
 * Response payload returned upon successful retrieval search.
 */
export interface RetrievalResponse {
  success: true;
  videoId: string;
  results: RetrievalResult[];
  requestId: string;
}

/**
 * Domain error codes specific to retrieval operations.
 */
export type RetrievalErrorCode =
  "RETRIEVAL_NOT_READY" | "INVALID_QUERY" | "INVALID_TOP_K";
